src/NHibernate/Transaction/AdoNetWithDistributedTransactionFactory.cs

```diff
@@ -24,52 +25,36 @@ public ITransaction CreateTransaction(ISessionImplementor session)
 
 		public void EnlistInDistributedTransactionIfNeeded(ISessionImplementor session)
 		{
+			if (!session.ConnectionManager.ShouldAutoJoinTransaction)
+			{
+				return;
+			}
+
+			// Ensure the session does not run on a thread supposed to be blocked, waiting
+			// for transaction completion.
+			session.TransactionContext?.WaitOne();
```

Comment:
Manually synchronizing end of second commit phase of session, since it may occur concurrently after scope disposal. `EnlistInDistributedTransactionIfNeeded` is called by most sensitive session operations, and I have added some `WaitOne` additional calls where needed.